src/client.rs

```diff
@@ -1912,31 +1954,107 @@ impl TryFrom<&ChallengeRef<'_>> for BearerChallenge {
+                "Cannot convert Docker-Content-Digest header value to string: {}",
+                e
+            )))
+        })
+}
+
+/// Given the optional digest header value and digest of the reference, returns the digest of the
+/// content, validating that the digest of the content matches the proper digest
+fn validate_digest(
+    body: &[u8],
+    digest_header: Option<String>,
+    reference_digest: Option<&str>,
+) -> Result<String> {
+    match compare_and_validate_digests(digest_header, reference_digest)? {
+        Some(digest) => calculate_and_validate(body, &digest),
+        // If we have neither, just digest the body
+        None => Ok(sha256_digest(body)),
+    }
+}
+
+/// This function compares the possible digest header value with a digest of a reference and returns
+/// the proper digest if it is valid. If both digests are provided, but they use different
+/// algorithms, then the header digest is returned as according to the spec it should be used for
+/// validation
+fn compare_and_validate_digests(
+    digest_header: Option<String>,
+    reference_digest: Option<&str>,
+) -> Result<Option<String>> {
+    match (digest_header, reference_digest) {
+        (None, None) => Ok(None),
+        (Some(digest), None) => Ok(Some(digest)),
+        (None, Some(digest)) => Ok(Some(digest.to_string())),
+        (Some(digest), Some(ref_digest)) => {
+            // According to the spec, it is just fine if the requested digest differs from the
+            // canonical digest from the headers if the algorithms are different. If they are the
+            // same, then the digests should match.
+            let (algo, _) = digest.split_once(':').ok_or_else(|| {
+                OciDistributionError::GenericError(Some(
+                    "Incorrect digest format on header value. Should have algorithm".to_string(),
+                ))
+            })?;
+            let (ref_algo, _) = ref_digest.split_once(':').ok_or_else(|| {
+                OciDistributionError::GenericError(Some(
+                    "Incorrect digest format on image reference. Should have algorithm".to_string(),
+                ))
+            })?;
+            if algo != ref_algo {
+                return Ok(Some(digest));
             }
+            if digest != ref_digest {
+                return Err(OciDistributionError::ContentDigestVerificationError(
+                    ContentDigestVerificationError {
+                        expected: ref_digest.to_string(),
+                        actual: digest,
+                    },
+                ));
+            }
+            Ok(Some(digest))
         }
-        Some(hv) => hv
-            .to_str()
-            .map(|s| s.to_string())
-            .map_err(|e| OciDistributionError::GenericError(Some(e.to_string()))),
     }
 }
 
+/// Helper for calculating and validating the digest of the given content
+fn calculate_and_validate(content: &[u8], digest: &str) -> Result<String> {
```

Comment:
It would be nice to have unit tests for these helper functions that have been introduced